protocols/messages.go

```diff
@@ -1,19 +1,98 @@
 package protocols
 
 import (
+	"bytes"
 	"encoding/json"
+	"sort"
 
+	"github.com/ethereum/go-ethereum/common"
 	"github.com/statechannels/go-nitro/channel/consensus_channel"
 	"github.com/statechannels/go-nitro/channel/state"
 	"github.com/statechannels/go-nitro/types"
 )
 
+const (
+	SignedStatePayload    PayloadType = "SignedStatePayload"
+	SignedProposalPayload PayloadType = "SignedProposalPayload"
+)
+
+type PayloadType string
+
 // Message is an object to be sent across the wire. It can contain a proposal and signed states, and is addressed to a counterparty.
 type Message struct {
-	To              types.Address
-	ObjectiveId     ObjectiveId
-	SignedStates    []state.SignedState
-	SignedProposals []consensus_channel.SignedProposal
+	To       types.Address
+	Payloads []MessagePayload
+}
+
+// MessagePayload is an objective id and EITHER a SignedState or SignedProposal.
+type MessagePayload struct {
+	ObjectiveId    ObjectiveId
+	SignedState    state.SignedState
+	SignedProposal consensus_channel.SignedProposal
+}
+
+// hasState returns true if the payload contains a signed state.
+func (p MessagePayload) hasState() bool {
+	stateHash, err := p.SignedState.State().Hash()
+	if err != nil {
+		return false
+	}
+
+	return bytes.Compare(stateHash.Bytes(), common.Hash{}.Bytes()) != 0
+}
+
+// hasProposal returns true if the payload contains a signed proposal.
+func (p MessagePayload) hasProposal() bool {
+	return p.SignedProposal.Proposal != consensus_channel.Proposal{}
+}
+
+// Type returns the type of the payload, either a SignedProposal or SignedState.
+func (p MessagePayload) Type() PayloadType {
+
+	if p.hasProposal() && !p.hasState() {
+		return SignedProposalPayload
+	} else if p.hasState() && !p.hasProposal() {
+		return SignedStatePayload
+	}
+	panic("payload has both state and proposal")
```

Comment:
That's a good point. I think I wrote this under the assumption that the it would be only dealing with a valid Payload.

We should probably add some kind of logic to validate a message so we don't have to rely on `Type()` generating an error. If we have that validation logic then I think it'd would be ok to leave the `panic` here, since there's no realistic scenario of that `panic` being called.